Date improvements

totalgee opened the pull request:
This is a fix for #3087.

- allow a new Date to be created without filling in all properties. In
      particular, `dayOfWeek` and `rawSeconds` will now be computed if they
      are not provided, and `hour`/`minute`/`second` now default to 0.
- add new `_ResolveDate` primitive to the language (corresponds to the
      new `prResolveDate` function), that resolves/corrects a
      partially-completed Date object.  It will calculate `dayOfWeek` and
      `rawSeconds`, and also "correct" other properties that may be out of
      their valid ranges (as a bonus, this feature can be used to perform simple
      date/time offsetting calculations).
- add Date comparison methods (`<` and `==`), that compare Dates based on
      their `rawSeconds` value.
- update documentation with new functionality.
- add a unit test for the Date class.

brianlheim commented:
thanks @totalgee!

a few initial questions:

why is it a failure condition that one of rawSeconds and dayOfWeek is specified but not the other? unless i'm missing something obvious, neither value constrains the other.

I would challenge the design choice of a hard error when both values have been specified. I think it would be more useful to treat it as a no-op in that case.

why have you defined some comparators but not all? I can't remember if the rest are inferred by defining lt and eq but in either case your documentation and tests only cover these two comparisons.

the primitive name is probably better as `_Date_Resolve` and the function as `prResolve`. `date` is superfluous in the latter since that is already the name of the class, but it is necessary for the primitive name, which will be in a global namespace. the `_Class_Function` style is used for other classes (on mobile so can't check but i'm pretty sure Array and String).

overall this is a really high quality submission. thank you!

totalgee commented:
Thanks, @brianlheim, for your comments!  I'm actually adding another new thing (string parsing to create new Date objects), and have also made some updates based on your comments (to be pushed soon).

> why is it a failure condition that one of rawSeconds and dayOfWeek is specified but not the other? unless i'm missing something obvious, neither value constrains the other.

I made it fail if either of them is missing (shouldn't break any existing code because currently it fails if you omit *any* of the 8 constructor values -- actually, the constructor doesn't fail but it will as soon as you try to display the Date object, which the post window will do for you pretty quickly).  I think maybe there was a change of behaviour at some point in the past, because the members are initialized to default values, but then the `*new` constructor overwrites them all.

I would really prefer that the constructor of Date didn't take either of these last two arguments, but didn't want to make a change that potentially might break existing code.  First, there is no good reason (I can think of) why the user should ever be expected to provide the `dayOfWeek`.  For `rawSeconds`, I think this should always be a computed value too.  You should either create a new Date using new (where you provide the YMD (and optional HMS) values and it calculates the weekday and epoch seconds), or use one of the other class methods that fill everything in with the current time, like `*localtime` or `*getDate` (not sure why we need both of these, they seem to give the same value ;-).  `*gmtime` is a bit different, so that's fine.  The difference is that those other constructor methods give you sub-second precision of the current time, whereas building a Date with the `*new` constructor will just give you single-second precision, similar to a filesystem date.

> I would challenge the design choice of a hard error when both values have been specified. I think it would be more useful to treat it as a no-op in that case.

There is no hard error when both values are specified.  In that case, we get the same (current) behaviour, where the user has filled in all the values themselves (and they may not be consistent, but so be it...).  I could change this to **always** do the "validation" when creating an object using `*new`.  What do you think?  But I'd have to decide which one "wins", if all the other arguments don't correspond with the `rawSeconds` value.  It really makes no sense to specify all these things!

As a side note (but related), there is a `Date.fromRawSeconds` method, added by the wslib Quark.  I left it alone for now.  It might make more sense as a native Date method (potentially in C++ code), but again I don't want to break anything or change too much here -- that Quark method seems to work fine with my new code (i.e. if I create a Date object `d`, then call `Date.fromRawSeconds(d.rawSeconds)`, we get an equivalent time).  Also, note that the underlying "rawSeconds" value *may* not match from platform to platform.  Most systems (all current SC platforms, I think?) use the "UNIX epoch" of Jan 1, 1970, but they aren't *obliged* to by the standard.

> why have you defined some comparators but not all? I can't remember if the rest are inferred by defining lt and eq but in either case your documentation and tests only cover these two comparisons.

Originally I'd defined them all, but then I looked at some other classes, and found several where only the '<' and '==' were defined, so I copied that style.  But no, the others are not inferred.  So I've gone and added them back in (<, <=, >, >=, ==)...   (-;

> the primitive name is probably better as _Date_Resolve and the function as prResolve [...]

Okay, I've fixed that, thanks for this more consistent naming.

totalgee commented:
Maybe I can change the `*new` method to remove the last two arguments.  I don't see anywhere it's used in the code in the SC repository.  It's possible people may be using it, but ... ?  It's not really useful as it is now...(IMHO)

Mostly it seems Date is currently just used to get the current time (and timestamp for files/logs), as well as for getting a time-based random seed.

brianlheim commented:
about to go to bed, but just wanted to say i appreciate your thoughtful response. also, i would advise against changing the method signature except as a last resort. you never know what bizarre use cases are out there.

totalgee commented:
Okay, I added a new commit, some new functionality and a bit of cleanup after @brianlheim's comments.  Did not change the `*new` method signature.

### Add string parsing to create Date objects
    
- add `Date.newFromFormattedString` class method (it uses a new
      `_Date_FromString` language primitive and underlying `prDateFromString`
      function). It allows you to specify a time/date string and parse it
      according to a second string, which uses the same formatting rules
      as the `format` instance method.
- make naming more consistent (`_ResolveDate` -> `_Date_Resolve`, based on
      code review feedback).
- add missing Date comparison methods (`<=`, `>`, `>=`).
- update docs and unit test.

totalgee commented:
Argh, it seems GCC < v5 doesn't implement std::get_time()...  Hence the build fail above.

```
/home/travis/build/supercollider/supercollider/lang/LangPrimSource/PyrUnixPrim.cpp:360:8: error: ‘get_time’ is not a member of ‘std’
 ss >> std::get_time(&tm0, formatString);
 ^
```

totalgee commented:
Okay, I changed the way the Date strings are parsed to use Boost, since GCC versions <5 (e.g. 4.8, which is used in the Travis build on Linux) doesn't implement C++11's `std::get_time()`, and Windows doesn't support an alternative function, `strptime()`).  Boost was a good cross-platform alternative.  In a few cases the Boost input parser's format flags are slightly different from the existing format stuff that uses strftime() (mostly there are just fewer valid options, I documented it with a link to the reference, and gave some examples in the help file).  I think this is "the good one"...  (-;

Question for @scztt (or anyone who knows): In order to activate my new Date unit tests for the Travis build, is it just a matter of adding an extra entry like the following, to `travis_test_run_proto.json`?

```
    {
      "test": "*",
      "suite": "TestDate"
    },
```

(My test class is in a new file: `testsuite/classlibrary/TestDate.sc`)
Thanks!

totalgee commented:
Okay, I tried what I asked above, but obviously I did something wrong with the unit tests and TravisCI, because now with the modified `travis_test_run_proto.json` it gives an error:

```
Starting virtual X frame buffer: Xvfb.
    [1/1] TestDate:* (0.0s)
           ! Encountered error: Class nil not found
           * completed without error
```

Do I need to do something to make sure the TestDate.sc file is parsed in the sclang tests (so the class is available)?  It's located in the same directory as all those other files that are tested, such as TestSignal.sc, TestArray or TestEnv.sc.  Any help? Thanks...

totalgee commented:
@scztt Okay, I experimented a bit further and managed to install qpm and run it locally (just by investigating the scripts in the .travis directory).  It seems to be checking out from https://github.com/supercollider-quarks/CommonTests (which claims to be deprecated, saying it should instead be using the quark in the core repository -- which is where I added my TestDate.sc file!)...

So -- should I submit a PR to add my new TestDate.sc file there?  A comment from @snappizz says: "We shouldn't really be editing this old deprecated repository, but unfortunately our CI system currently uses it, so this modification serves as a quick fix."  Or should I just remove my test again from `travis_test_run_proto.json`?  Sorry to be a nuisance...thanks!

brianlheim commented on lang/LangPrimSource/PyrUnixPrim.cpp:
```diff
@@ -341,6 +342,80 @@ int prGMTime(struct VMGlobals *g, int numArgsPushed)
+	if (!IsNil(slots+6) || !IsNil(slots+7)) {
+		error("dayOfWeek or rawSeconds are already defined\n");
+		return errFailed;
+	}
```
i think i may have been unclear before.

why is this a failure condition? as opposed to filling in dayOfWeek if missing and rawSeconds if missing, and returning without error as long as type checks pass?

totalgee replied:
Well, I'd put some logic in `Date.new` that only calls "resolve" if one of those two last arguments is Nil (see Date.sc).  To be "conservative" I put the above code in prDateResolve() that ensures we're being called under those (expected) conditions (one of them Nil).  But as you suggest, maybe this is too extreme.  I guess I'll just remove this check in the C++ code.  The code that follows will work correctly, filling those properties in (regardless of whether or not they were Nil, they will be overwritten -- the incoming values aren't saved).  But I think that's okay, and sounds like you do, too.  Thanks.

totalgee replied:
To be clear, with the new code (where these checks are removed), if I do one of these:

```supercollider
Date(1977, 5, 25, 13, 30, 22, dayOfWeek: 2);
// -> Wed May 25 13:30:22 1977
```

Even if the incoming `dayOfWeek` (2) is passed in wrong, the new method will return a valid (corrected) date (`Wed May 25 13:30:22 1977`, where the valid dayOfWeek (Wednesday) is 3):

```supercollider
Date(1977, 5, 25, 13, 30, 22, rawSeconds: 1234567.8);
// -> Wed May 25 13:30:22 1977
```

Even if the `rawSeconds` is wrong here, a valid date will be returned (same as above).  This "valid" Date will have a computed `rawSeconds` of 233407822, not the 1234567.8 the user passed into new.  The other option in the case of passing in rawSeconds would be to ignore all the other arguments (year/month/day/etc) and give priority to the rawSeconds, and fill in all the rest...  That's easy to do too, if you think it's better.  Personally, I think if you specify all the other params, it's most likely you want them to "rule" (this is why I suggested it would be better if neither dayOfWeek nor rawSeconds were arguments to `new`)...

brianlheim commented:
@totalgee w/r/t tests, no need to apologize, it's a mess on our end right now admittedly. i'll bring it up at the dev meeting today (in an hour) and let you know how we can best proceed. i think what you're doing is on the right track, thanks for putting in that effort :)

brianlheim commented on lang/LangPrimSource/PyrUnixPrim.cpp:
```diff
@@ -341,6 +342,80 @@ int prGMTime(struct VMGlobals *g, int numArgsPushed)
+	if (IsNil(slots + 0)) {
+		SetNil(a);
+		return errNone;
+	}
```
what is the purpose of this block?

totalgee replied:
This was another attempt at being "conservative".  (-;  If certain of the incoming arguments are Nil, then we won't be able to resolve a date/time (unless we decide on some defaults for that case -- what would be the "default" year/month/day?).  Instead of checking if *any* of the required inputs was Nil (as most typically this is what would happen e.g. if you called `Date.new` with no arguments), I just checked the first argument here (if it's Nil, it sets the function's return value back to scripting as Nil, rather than returning an error...).  But I don't know that this was a good thing to do...what do you suggest for this case?  Return errFailed instead?  Or use some kind of default values?

totalgee replied:
Here's what I would propose.

1. Have default values for the first six arguments (Y,M,D,H,M,S) in `Date.new`, so if you call Date.new it would create *something* valid and not give an error (as it does with my current changes, and did *before* my changes!).  For everything except `year`, there is an obvious and reasonable default (month=1, day=1, hour=0, minute=0, second=0).  For year, we could just arbitrarily choose something.  In the Date class, someone decided to initialize the year member to 2000, although the existing `new` method always overwrites what's there, so there effectively is no default date.  So if you do `Date.new` right now, you get an error, because all the members are set to Nil, and then the asString method fails when the post window tries to output an "empty" Date object.  But the year could arbitrarily default to 2000, or 2017, or have `Date.new` give the current time (`Date.localtime`), or...?  As I mentioned in an earlier comment, I'd really prefer that the `new` method only have Y, M, D, H, M, S arguments, not `dayOfWeek` or `rawSeconds`.  Especially not to have `dayOfWeek`, although I understand the desire to leave it as-is, because it would be a breaking API change.

```supercollider
Date.new;
// Current behaviour: ERROR: Primitive '_ArrayCat' failed.
// Proposed: -> Sat Jan  1 00:00:00 2000 // add defaults for all parameters (e.g. year 2000, or ...?)

Date(2017);
// Current behaviour: ERROR: Primitive '_Date_Resolve' failed. (because month and day don't have default values)
// Proposed: -> Sun Jan  1 00:00:00 2017 // month/day default to 1, hour/min/sec to 0
```

2. If you choose to set `rawSeconds`, it would always override/recalculate all the other values, to remain internally consistent (they are two views on the same data).  Would be true both if you set the `rawSeconds` argument to `new`, but also if you did `d.rawSeconds = <something>` on an existing Date object.

```supercollider
Date(rawSeconds: -14186520);
// Current behaviour: ERROR: Primitive '_ArrayCat' failed.
// Proposed: -> Sun Jul 20 20:18:00 1969 // compute all the other fields from the rawSeconds

// The following would use the rawSeconds value, ignoring the rest of the args since they're inconsistent
Date(2017, 12, 30, 5, 13, 22, rawSeconds: 1234567890);
// Current behaviour (with my changes): -> Sat Dec 30 05:13:22 2017 // ignore rawSeconds
// Proposed: -> Sat Feb 15 00:31:30 2009 // N.B. *not* December 30, 2017! Use rawSeconds, ignore the rest

d = Date(2005, 5, 31);
// Current behaviour: ERROR: Primitive '_Date_Resolve' failed.
// Proposed (implemented in this PR): -> Tue May 31 00:00:00 2005
d.rawSeconds = d.rawSeconds + 10000; // add 10k seconds, which will update the other fields
// Current behaviour: will add 10000 to a valid Date's rawSeconds, but won't modify any other fields (like hour/minute/second, so the Date will now be "inconsistent"
// Proposed: -> Tue Jun  1 01:46:40 2005 // update other fields to match new rawSeconds
```

3. Note that rawSeconds is normally a Double (and will indeed be fractional, if you use methods like `Date.localtime` to get the high-resolution current time).  Seconds is currently always an Int (generally 0-59).  I wouldn't want to modify this, because it would affect the code that outputs Date strings.  Just to be aware, though: if you have a high-precision Date (e.g. from Date.localtime), and then create a new Date using just the computed YMD/HMS values, you won't get an equal Date (because mtkime only resolves seconds).  I could make it add in the factional bit to the rawSeconds, but right now it doesn't...

```supercollider
d = Date.localtime;
[d.asString, d.rawSeconds].postln;
// -> [ Sun Oct 29 18:07:03 2017, 1509296823.8734 ]

e = Date(d.year, d.month, d.day, d.hour, d.minute, d.second);
[e.asString, e.rawSeconds].postln;
// -> [ Sun Oct 29 18:07:03 2017, 1509296823 ]
d == e;
// -> false

// However, if I implement what I propose above, you could do:
e = Date(rawSeconds: d.rawSeconds);
// The you would get
d == e;
// -> true
```

Also, note that `d.second` for the above example (also for `e`) returns `3`, not `3.8734` (which it *might*, but doesn't!).  I don't think I'd like to change this (this is much more likely to break existing code), although there certainly could be an argument made for that behaviour (fractional seconds) to make sense!

4. One last thing.  With my current changes, if you pass the constructor an invalid date (like day-of-month 60, or hour 200), it will "correct" it to the equivalent valid date.  This is quite cool and useful, see my examples in the documentation or unit tests, to use it to calculate relative dates.  However, I don't do any validation (right now) if you modify the properties individually, once you have an existing Date object.  I could, though, always do revalidation.  So, for example, if you did:

```supercollider
d = Date(2005, 5, 31);
// -> Tue May 31 00:00:00 2005

// You could offset a number of months or days (or years, hours, minutes, seconds) in the future or past quite easily:
d.month = d.month + 20;
// -> Wed Jan 31 00:00:00 2007

d.hour = -50; // 50 hours earlier
// -> Sun Jan 28 22:00:00 2007
```

Anyhow, that's it for now...  What do you think?  Should I implement at least the changes in those first two points?  Thanks.

brianlheim replied:
These are all essentially changes to behavior of the constructor except for (4), right?

I think I understand where you're going now with this. Let me state the spec for the constructor a little more concisely:

1. If all args nil -> return "now" (or all-defaults, but I personally find that less useful)
2. Else, if rawSeconds is provided, use it to set (YMD:hms) regardless of whether they are nil or not.
3. Else, if year is not provided, throw an error (can't fill in defaults)
4. Else, fill in defaults for any of (MD:hms) that are nil, and set rawSeconds from (YMD:hms).
5. Set dayOfWeek, overwriting any existing value

Is this correct? From my understanding dayOfWeek is always set internally, possibly changing the user-provided value if it's different.

I think you should provide the resolution operation as a user-facing utility. In other words, leave the current behavior for property setters as-is, but also give the ability to call `resolve` to turn it into a valid date again. You might have to provide two - one to resolve from `rawSeconds`, and one to resolve from other fields. Or, have `resolve` always use the other fields, and rely on the constructor for the second case.

I think paying for what you use is more sensible here, since it's fairly expensive to recompute all the fields on every set of minute or second.

```supercollider
d.hour = d.hour - 1;
d.month = d.month + 3;
d.minute = d.minute + 100;
d.resolve;

d.rawSeconds = d.rawSeconds + 3.5;
d.resolveFromRaw; // first idea

d = Date(d.rawSeconds + 3.5); // second idea
```

Re point (3), I think leave it as is. Currently, you can get millis with `d.rawSeconds % 1`, and making `d.seconds` return a rounded value, instead of truncated, would fudge that. And besides as you point out, it would probably not be very kind to suddenly change the contract of such a low-level and subtle feature.

still not convinced we've pinned down the best solution, but this is getting closer I think!

totalgee replied:
That sounds good to me.  For *your* point 3), I guess you mean "if year is nil but some of (MD:hms) are non-nil, throw an error"?  I'm just noting that, because up to that point, I'd been planning to provide the defaults for MD:hms in scripting:

```supercollider
*new { arg year, month=1, day=1, hour=0, minute=0, second=0, dayOfWeek, rawSeconds;
```

But then if you call Date.new you'll get Y:nil and the rest with default values, so your point 1) would be hard to detect (unless the user wrote `Date(nil, nil, nil, nil, nil, nil)`, which would be pretty annoying! ;-).  So I'll have to have no defaults in the `*new` arguments declaration.

Quick summary with code examples (ready for unit tests!):

```supercollider
Date.new; // current local time
// -> Mon Oct 30 20:25:35 2017
Date(rawSeconds: 1234567890); // init from rawSeconds
// -> Sat Feb 15 00:31:30 2009
Date(1990, 1, 1, 0, 0, 0, rawSeconds: 1234567890); // rawSeconds overrides others
// -> Sat Feb 15 00:31:30 2009
Date(month: 3); // year or rawSeconds must be provided
Date(nil, 1, 1, 0, 0, 0) // ditto: year can't be nil
// ERROR: year must be specified
Date(1990); // defaults exist for MD:hms
// -> Mon Jan  1 00:00:00 1990
Date(1990, 2, 1, dayOfWeek: 1); // dayOfWeek (if we keep it as an arg) is always overwritten
// -> Thu Feb  1 00:00:00 1990
Date(dayOfWeek: 1, rawSeconds: 633826800); // same if you init using rawSeconds, dayOfWeek overwritten
// -> Thu Feb  1 00:00:00 1990
```

Just to be clear on another point... you're still convinced I should keep the Date.new signature the same, with dayOfWeek, even if it's totally useless (just to maintain "backward compatibility")?

Thanks again for this back and forth, it's very helpful!

totalgee replied:
BTW, I'm fine with your `resolve` and `resolveFromRaw` suggestion; I'd thought of maybe doing something similar, that maybe it was too "heavy" to resolve all the time -- and this also allows several values to be modified before calculating the "resolve", as you say.

brianlheim replied:
> you're still convinced I should keep the Date.new signature the same, with dayOfWeek, even if it's totally useless (just to maintain "backward compatibility")?

Yes, it's not really that bad to keep it in, although i understand your annoyance :)

> BTW, I'm fine with your resolve and resolveFromRaw suggestion; I'd thought of maybe doing something similar, that maybe it was too "heavy" to resolve all the time

++ what do you think about leaving out `resolveFromRaw` in favor of using the constructor? I think it would be about the same performance-wise and I don't see much of a reason to add a second way to do it beyond it potentially being faster.

> I guess you mean "if year is nil but some of (MD:hms) are non-nil, throw an error"?

Yes -- it's in an if-else branch, the first `if` of which is taken if all of (YMD:hms) are nil. So we only get to point (3) if one of those is non-nil. Sorry if that was confusing

> Thanks again for this back and forth, it's very helpful!

Same to you :)

totalgee replied:
> ++ what do you think about leaving out resolveFromRaw in favor of using the constructor? I think it would be about the same performance-wise and I don't see much of a reason to add a second way to do it beyond it potentially being faster.

That's fine, totally agree, will do that.

So I'll go ahead and implement the last of these details we've talked about...

My last remaining "unknown", then, is about the unit testing stuff.  I'll write my unit tests.  But about the Travis automation -- any resolution/clarity on that?  Or should I just take out the bit I added (hte lines in `travis_test_run_proto.json`)?  If not, I'll need some guidance on how to enable Travis testing properly, and whether I'm on the right track by adding it to the (deprecated) CommonTests repo...or not... (I understand that if so, it's a temporary solution)

Thanks.

brianlheim replied:
> But about the Travis automation -- any resolution/clarity on that? Or should I just take out the bit I added (hte lines in travis_test_run_proto.json)? If not, I'll need some guidance on how to enable Travis testing properly, and whether I'm on the right track by adding it to the (deprecated) CommonTests repo...or not... (I understand that if so, it's a temporary solution)

For the sake of us all staying healthy, I think let's just go with the quick way for now - you've done it right with the lines you've added here, and adding it to CommonTests is the way to go. But yes, hopefully very temporary. Sorry again for the confusion.

totalgee commented:
Okay, I've put in the "final" changes.  I added a lot of tests, and I think the added/changed documentation is quite clear (relevant parts copied below for easier review; though some formatting missing here compared to SCDoc).

-----

## Class Methods

### Date.new(year, month, day, hour, minute, second, dayOfWeek, rawSeconds)
Create a new Date with the given properties (all are Integer values, except rawSeconds, which is a Float). These arguments are also the names of instance variables of the Date object.

The constructor offers several possibilities, which are evaluated in the following order of priority:

1. If Date.new is called without any arguments, it will return the **current time**, the same as calling *localtime.
```supercollider
// The following are equivalent, they return "now" (at the time of call):
Date.new;
Date.localtime;
```

2. If a rawSeconds argument is specified, it is used to set (overwrite) all the other properties (year, month, day, hour, minute, second, dayOfWeek), regardless of whether they are nil or not.
```supercollider
(
var d = Date.new.postln;
var e = Date(rawSeconds: d.rawSeconds); // create a new Date with the same time
d == e; // true
)

Date(2000, 1, 1, rawSeconds: 1234567890) != Date(2000, 1, 1); // true
Date(2000, 1, 1, rawSeconds: 1234567890) == Date(rawSeconds: 1234567890); // true
```

3. Most properties have default values (month=1, day=1, hour=0, minute=0, second=0). If no year is provided, an error is thrown (there is no default year). When rawSeconds is unspecified, it is computed based on the resulting date/time.
```supercollider
Date(1969, 7, 20, 20, 18); // -> Sun Jul 20 20:18:00 1969
Date(2016); // -> Fri Jan  1 00:00:00 2016
Date(2010, day: 15, minute: 30); // -> Fri Jan 15 00:30:00 2010
try { Date(month: 7) } { |err| err.postln }; // ERROR: year must be a valid Integer
(
var d = Date(2017, 10, 25);
[d.asString, d.dayOfWeek, d.rawSeconds]; // dayOfWeek and rawSeconds were computed
)
```

4. The dayOfWeek argument is **always calculated** and if it is specified explicitly, the value will be overwritten (it only remains a constructor argument for reasons of backward compatibility).
```supercollider
// There is no point (but no harm) in specifying the dayOfWeek argument,
// it will always be calculated:
Date(2017, 10, 31, dayOfWeek: 0).dayOfWeek; // calculated to be 2 (Tuesday), not 0 (Sunday)
"Party like it's" + Date(1999, 12, 31, 23, 59, 59);
```

It is possible to specify dates with "out of range" values that will be resolved into a valid range -- this allows simple date offset calculations to be performed (see third example below).
```supercollider
(
// Simple date/time offsetting can be performed:
("One hundred days before Christmas:" + Date(2010, 12, 25 - 100).format("%B %d")).postln;

// Invalid dates are "corrected"
Date(2017, 2, 29).postln; // March 1: 2017 was not a leap year!
)
```

### Date.newFromFormattedString(string, format)
Create a new Date object from a date/time string, which is parsed according to the provided format string.

```supercollider
Date.newFromFormattedString("2017-10-25 [13:25:55]", "%Y-%m-%d [%H:%M:%S]");
Date.newFromFormattedString("1981.9.30 @ 17:33", "%Y.%m.%d @ %H:%M");

(
// Round-trip via string, using "stamp"
var stampString = Date(2013, 5, 25, 10, 9, 8).stamp;
var d = Date.newFromFormattedString(stampString, "%y%m%d_%H%M%S");
(stampString + "-->" + d).postln;
)
```

**Arguments:**
- **string** The date/time string to parse, for example "2017-10-25 13:25:55".
- **format** The format to use to parse the date string. Note that the supported input formats are slightly different from those of the -format (output) method. In particular, some of the "combined" formats (such as %F or %T) are not supported for input, you must expand them out ("%F" -> "%Y-%m-%d" and "%T" -> "%H:%M:%S"). The full list of supported input format flags may be found [here](http://www.boost.org/doc/libs/1_63_0/doc/html/date_time/date_time_io.html).

  To parse the above example string, you would use "%Y-%m-%d %H:%M:%S".

## Instance Methods

### .resolve
Resolve any inconsistencies in the Date object, updating all the fields (in particular dayOfWeek and rawSeconds) as necessary.

If the instance properties (year, month, day, hour, minute, second) are modified, the Date will be in an inconsistent state -- its rawSeconds won't reflect the modifications, and the properties may fall outside of "normal" ranges (month: 1-12, day:1-31, hour: 0-23, minute/second: 0-59). Also, using the string conversion methods may give strange results. Calling the resolve method will put things back into a consistent state.

```supercollider
// Simple date/time offsetting can be performed:
d = Date(2017, 9, 30, 10, 30); // -> Sat Sep 30 10:30:00 2017
[d.dayOfWeek, d.rawSeconds]; // -> [ 6, 1506760200 ]
d.hour = d.hour - 13; // -> Sat Sep 30 -03:30:00 2017
d.month = d.month + 4; // -> Sat ??? 30 -03:30:00 2017
d.minute = d.minute + 200; // -> Sat ??? 30 -03:230:00 2017
d.rawSeconds; // hasn't changed: still 1506760200

// Now resolve everything back into reasonable ranges, applying the offsets:
d.resolve; // -> Tue Jan 30 00:50:00 2018
[d.dayOfWeek, d.rawSeconds]; // -> [ 2, 1517269800 ]
```

Note that there is no "resolveFromRawSeconds" method. If you modify the rawSeconds property, the other properties **won't** change to reflect it. If you want to "resolve" in the other direction, simply create a new Date object using the rawSeconds:

```supercollider
d = Date(1999, 12, 31); // -> Fri Dec 31 00:00:00 1999
d.rawSeconds; // -> 946594800

// Add 1 million seconds (~11.6 days)
d.rawSeconds = d.rawSeconds + 1e6; // still shows -> Fri Dec 31 00:00:00 1999
d.rawSeconds; // -> 947594800
// rawSeconds different but the other properties haven't changed

// If we were to call d.resolve, then rawSeconds would be reset to match all
// the other properties.  So in this case, where we want to "resolve" from the
// rawSeconds, just create a new Date object:
d = Date(rawSeconds: d.rawSeconds);
// -> Tue Jan 11 13:46:40 2000
```

### <(other)
### <=(other)
### >(other)
### >=(other)
Compare two Dates based on their rawSeconds.
```supercollider
Date(1999, 12, 31, 23, 59, 59) < Date(2000, 1, 1, 0, 0, 0);
Date(1999, 12, 31, 23, 59, 59) <= Date(2000, 1, 1, 0, 0, -1);
Date(2005, 1, 1) > Date(2004, 1, 2);
Date(2005, 2, 13) >= Date(2005, 2, 13);
```

### ==(other)
### !=(other)
Test whether two Dates are equal/unequal, based on their rawSeconds.
```supercollider
Date(2016, 11, 4) == Date(2016, 10, 4 + 31);
Date(2016) != Date(2017);
```
-----

  If someone could please merge the changes in CommonTests (https://github.com/supercollider-quarks/CommonTests/pull/9) when this gets merged into SC...  It's kind of a chicken-and-egg problem, I need both repository changes at once, in order for the Travis tests to pass, presumably!

brianlheim commented:
Ok, I've merged the PR to CommonTests. Re-running some builds to make sure that didn't break anything (it shouldn't). Thanks again!

Without having time to thoroughly review what you've written in the help file, my first piece of advice would be to shorten it up if you can. I think in its current form, most people will not bother to read the whole thing. The logic seems harder to understand than it really is. For one, you say "evaluated in this order of priority," but part (4) always happens and parts (1-3) are mutually exclusive behaviours.

Also - sorry for not noticing this before; would you consider `fromString` instead of `newFromFormattedString`? This would match a similar paradigm in `fromArray` (used in Color, Rect, 2DArray, etc). I guess we also use `newFromX` in places too. Mainly, `formatted` is a superfluous word in this method name, IMO. (What would `newFromUnformattedString` look like?)

totalgee commented:
> Ok, I've merged the PR to CommonTests. Re-running some builds to make sure that didn't break anything (it shouldn't). Thanks again!

Thanks, sorry, but could you please merge it again (new PR https://github.com/supercollider-quarks/CommonTests/pull/10)?  Had to update the unit tests again...

> Without having time to thoroughly review what you've written in the help file, my first piece of advice would be to shorten it up if you can. I think in its current form, most people will not bother to read the whole thing. The logic seems harder to understand than it really is. For one, you say "evaluated in this order of priority," but part (4) always happens and parts (1-3) are mutually exclusive behaviours.

I tried to make the \*new logic a bit simpler, and now only has the 3 real cases... also reworded a bit and got rid of some of the extraneous examples, which were chopping it up a bit more.  I also moved some things into "note::" breakout sections, so they stand out more.

I get what you're saying about shortening, but I think it's very easy to err on the side of too *little* documentation and examples (a historical problem in SC)...  I think what I've written now is pretty clear and useful, but maybe only to me...  (-;

Any other specific suggestions are welcome, but I think it's already way better than it was originally.  Especially that there are concrete usage examples for all the things I added -- I understand that people may not read *all* the text, but they can skim those examples, and can surely find what they want to do.  (and the text adds some helpful bits that will be useful if people run into trouble)

> Also - sorry for not noticing this before; would you consider fromString instead of newFromFormattedString? This would match a similar paradigm in fromArray (used in Color, Rect, 2DArray, etc). I guess we also use newFromX in places too. Mainly, formatted is a superfluous word in this method name, IMO. (What would newFromUnformattedString look like?)

Fixed that, now it's `Date.fromString` (tests and doc updated).

I'm going to have to leave this for now, it's taken way more time than I'd expected.  If there are any more things, let me know and I'll try to get to them later this week.

Thanks again.

telephon commented on SCClassLibrary/Common/Math/Date.sc:
```diff
@@ -5,12 +5,24 @@ Date {
 			rawSeconds);
-		if (dayOfWeek.isNil or: {rawSeconds.isNil}) {
-			instance.prResolve;
+		case
```
somehow, I'd prefer if `resolve` would take care of all those case distinctions. Then they wouln't have to clutter the `*new` implementation.

totalgee replied:
Do you mean just moving this logic into the Date class `resolve` instance method, or into the C++ helper function?

totalgee replied:
Actually, I wanted `resolve` to always resolve from YMD:HMS -> rawSeconds (there is currently a private helper `prResolveFromRawSeconds` to go the other way, rawSeconds -> YMD:HMS).  In discussion with @brianlheim we agreed that when one needed to convert *from* rawSeconds, it was acceptable to create a new Date instance, passing only the rawSeconds argument.

So if I move the current `*new` logic into `resolve`, then it won't work properly when you call `resolve` on an existing Date object (which will have a filled-in rawSeconds variable).  In this case, you'd want it to recalculate rawSeconds from YMD:HMS, but it would do the opposite.  So I want to be able to distinguish between creating a new Date vs requesting `resolve` on an existing instance.

However, I guess we **could** require the user to manually "nil" out the rawSeconds field (on an existing Date instance) if they want this behaviour, then we could have it all be done by `resolve`.  If you prefer that, I would change it so that:

1. When creating a new instance (and calling `resolve` on an existing instance), it would use the rawSeconds field if provided, overwriting everything else.
2. Calling `resolve` on an existing instance would "resolve from rawSeconds" (rawSeconds -> YMD:HMS).  e.g. If you did: `d.year = d.year+1; d.resolve;` you would lose your modified year and go back to the original date, because everything would be recomputed from rawSeconds.
3. When you want to do the reverse (YMD:HMS -> rawSeconds), you would have to do: `d.year = d.year+1; d.rawSeconds = nil; d.resolve;`

I'm not sure this is the best idea, though, because my intuition tells me one will more often want to manipulate Dates in the YMD:HMS format, and the rawSeconds would be mainly used internally (for comparing Dates, etc).

Hmm... An alternative approach would be to have `*new` call `prResolveFromRawSeconds` instead, which would be modified to resolve from rawSeconds if it exists, but otherwise perform regular `resolve` (and return `localtime` if everything is unspecified).  Then the current logic of the `resolve` method could be left alone, and `*new` would also get rid of this cluttery logic you don't like.  How's that sound?

totalgee replied:
In the end it makes more sense to just add an initialization helper method.  I did that, and updated my terminology to use the correct term ("instance variables").  See [here](https://github.com/supercollider/supercollider/pull/3259/commits/30644b0).

telephon commented on lang/LangPrimSource/PyrUnixPrim.cpp:
```diff
@@ -305,18 +305,31 @@ int prUnix_Errno(struct VMGlobals *g, int numArgsPushed)
-	SetFloat(slots+7, std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count() * 1.0e-9);
+}
+
+// Set all Date members (including rawSeconds)
```
just a comment: in sclang, "member variables" are usually called "instance variables"

totalgee replied:
Okay, thanks, updating all uses of "member" to "instance variables".

brianlheim commented:
@totalgee - I haven't forgotten this! Will try to get around to looking at it soon. Thanks again for your work. I will make sure to offer more specific suggestions on the documentation if I have any.

Merged your PR to CommonTests. When this is ready to be finalized, the tests should also be put in `testsuite/classlibrary` in this repo.

totalgee commented:
@brianlheim - Thanks.  I made a few last changes (one more commit) based on @telephon's comments.

Also, don't worry, the test code (that you merged into CommonTests, thanks!) is already in `testsuite/classlibrary` (in this PR).

totalgee commented:
By the way, should I be merging this PR with `develop`?  I think right now I based it off `master`.  It should auto-merge just fine (there is no overlap with anything that's been done on `develop` since September), but let me know and I'll get this PR up to date with `develop`.

brianlheim commented:
@totalgee Thanks! Yes, this PR will go into develop. It's already pointed to merge at develop and it looks like there are no merge issues, so nothing more needed on your end.

patrickdupuis commented:
If this PR is against `develop`, then it should be milistoned as `3.10`.

brianlheim commented:
good point, I agree

brianlheim commented:
@totalgee - the appveyor integration revealed an important problem with your branch, which is that constexpr is not supported by the VS 2013 compiler.

totalgee commented:
Okay, removed the constexpr, now it compiles and passes all tests.  (BTW, any plans to update to VS2015, at least?)

brianlheim commented:
We're pinned to 2013 because of a dependency on an old version of Qt. Upgrade should come before 3.10, possibly to 2017.

brianlheim commented:
Coming back to this now (I'd really like to wrap up discussion and merge it ASAP). I'd like to suggest the following API changes. Two main goals:

1. making sure the object always stays in an consistent state. In general, this way of designing objects is easier to reason about and to use; programming errors can be caught earlier as well.
2. tighten the interface and deprecate elements that are confusing or redundant

- constructor:
  - throw error if `rawSeconds` and one of YMD:hms/day-of-week is specified
  - warn if day-of-week is used. We ought to deprecate this and remove it in a major version bump. Almost all usage should look like one of `Date(yr, month, day, hr, min, sec)` or `Date(rawSeconds: r)`, so removing this argument should actually be painless for most people, unless (1) they are specifying day of week, in which case they're already receiving warnings, or (2) they are specifying raw seconds as `Date(nil,nil,nil,nil,nil,nil,nil, r)` (hopefully not), or (3) they are explicitly specifying `Date(dayOfWeek: nil)`, which is easy to correct.
- rawSeconds_:
  - auto-resolve on set (add a prResolveFromRaw and rename resolve to prResolveFromDateFields)
- resolve:
  - keep private (-> prResolveFromDateFields)
- setting year, month, day, hour, minute, second:
  - prResolveFromDateFields
- setting day of week:
  - disallow, and mark deprecated
- Date.getDate:
  - deprecate, same as Date.localtime
- remove the default values from `Date { var <>year=2000, ..., <>dayOfWeek=6, <>rawSeconds=0`, since these are wrong now.

I would also recommend throwing an error for `Date()` instead of returning localtime -- there is no obvious 'default' `Date` object (it could be unix epoch, now, Jan 1st 1 CE), so providing this shortcut for getting the current time is only going to confuse people. (In general: don't allow default construction of objects that have no clear default-constructed state.) `localtime` is long, I'd consider adding `Date.now` to alias it (and possibly deprecating `localtime` in the process, although this is useful to contrast with `gmtime`).

Also want to note that I am willing to implement the rest of these changes and polish the code if there's anything I notice. @totalgee in private communication has given me the go-ahead to take over this PR if he doesn't have time to work on it.

totalgee commented:
Thanks, @brianlheim - indeed at the moment I'm unfortunately unable to work on this, so if it needs to go in soon, it would be appreciated if you could take over any last few tweaks. What you propose above makes sense to me; thanks for reviewing it. I had indeed gone for the "don't break anything" approach -- but your suggestions to deprecate certain methods and throw/post warnings for undesired usages (such as the user specifying the day of week) are better.

beatboxchad commented:
Hey @brianlheim I've just finished rebasing this branch onto latest develop. This looks like a simple, unambiguous piece of grunt work where y'all have already done most of the hard thinking. I'd love to drive it home, and will likely have quick questions for you during the week. Is there anything I should know before continuing to dive in?

brianlheim commented:
@beatboxchad  what do you mean by "drive it home"?

beatboxchad commented:
I wanted to implement the small changes you suggested (reviewing the branch
you had in case you'd already done so). If I ended up making any changes, I
was going to ask whether I should push to the existing branch upon gaining
access, or make a fresh PR, or what.

Tl;dr I wanted to perform any work necessary to get this PR closed and its
changes into develop.

Chad

On Sun, Dec 15, 2019, 1:24 PM Brian Heim <notifications@github.com> wrote:

> @beatboxchad <https://github.com/beatboxchad> what do you mean by "drive
> it home"?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/supercollider/supercollider/pull/3259?email_source=notifications&email_token=AARIDN32AA7TJDLS3WGUDSTQY2AABA5CNFSM4EAXTP7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG5AGOQ#issuecomment-565838650>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARIDN5PWE5UH2IDCJAIQ4DQY2AABANCNFSM4EAXTP7A>
> .
>

brianlheim commented:
ah okay. yes please, go ahead! if the merge conflicts look unambiguous to resolve. you will have to open a separate PR, as you can't force push to totalgee's branch.

beatboxchad commented:
I didn't get to work on this during the week, but have completed a local rebase. Next I will make sure all the suggested changes have been implemented, make any minor corrections, and open a new PR.

brianlheim commented:
@beatboxchad hey! any update on this? i would like to finish it up, but if you've got a branch lying around please let me know!

totalgee commented:
Hi @beatboxchad , I really appreciate you offering to tidy up the final details (and merging, etc) of this very old PR. Did you manage to finish that work? It would be nice to get this finally merged (or not, though it would be a pity to throw away this work from 2017 ;-). I don't know what is your latest state of things, or where to find it. Please let us know if you can. Thanks!

telephon commented on SCClassLibrary/Common/Math/Date.sc:
```diff
@@ -74,4 +107,22 @@ Date {
+	>= { arg other;
+		^this.rawSeconds >= other.rawSeconds
+	}
+	== { arg other;
```
If we have an `==` method, we also need a `hash`method.
```
hash {
^this.instVarHash([\rawSeconds])
}
```

totalgee replied:
Okay, added this...thanks!

telephon reviewed (commented):
just one little change, the rest looks good!

totalgee commented:
I'm going back and merging my branch with the latest `develop` branch. Just testing it now. If someone can remind (tell) me how to run the SC tests locally, that would be helpful. Does `make test` (after doing a CMake build) cover everything, or are there more tests I should run? Mainly I want to make sure the things in `testsuite/classlibrary` are run.

totalgee commented:
I see some of the formatting changed in the intervening years. What is the way to locally get this into shape so it passes the linter? I'm trying to follow the instructions [here](https://github.com/supercollider/supercollider/wiki/Cpp-formatting-instructions) but so far have had no luck (I'm using clang-10 rather than clang-8, and didn't have luck going back to that older version so far)...

totalgee commented:
Ah, I managed to figure it out. I got the right version set up, but then I needed to use `formatall` as an argument.

totalgee commented:
Okay, this is merged with `develop` and ready to go (Date unit tests still pass), not including the changes suggested [here](https://github.com/supercollider/supercollider/pull/3259#issuecomment-466680103) (perhaps those can be done in a later phase if they're deemed necessary).

Thanks.

telephon reviewed (approved):
OK, all works fine.

I wondered whether `Date` should be a subclass of `Magnitude`, but well, not necessary. Just if it is fun, this would make sense.

dyfer reviewed (commented):
Thanks for reviving this!

dyfer commented:
Thanks @totalgee !
@telephon is this still good to go?

telephon commented:
@dyfer yes, this is still good. Can't wait for the date!

dyfer commented:
Sounds good, merging it `Date.localtime`.

dyfer commented:
Also, should we keep the `API change` label here? I don't think any functionality gets broken by this change, correct?

totalgee commented:
Thank you so much, I'm glad these old changes finally got merged in and weren't lost to the sands of time!

One question (I guess for @telephon or @dyfer?): do you still want any of the changes suggested in this Moss comment?

https://github.com/supercollider/supercollider/pull/3259#issuecomment-466680103

In particular (edited to only include what was not done):

> * constructor:
>   - throw error if rawSeconds and one of YMD:hms/day-of-week is specified
>   - warn if day-of-week is used. We ought to deprecate this and remove it in a major version bump. Almost all usage should look like one of Date(yr, month, day, hr, min, sec) or Date(rawSeconds: r), so removing this argument should actually be painless for most people, unless (1) they are specifying day of week, in which case they're already receiving warnings, or (2) they are specifying raw seconds as Date(nil,nil,nil,nil,nil,nil,nil, r) (hopefully not), or (3) they are explicitly specifying Date(dayOfWeek: nil), which is easy to correct.
> * rawSeconds_:
>   - auto-resolve on set (add a prResolveFromRaw and rename resolve to prResolveFromDateFields)
> * resolve:
>   - keep private (-> prResolveFromDateFields)
> * setting year, month, day, hour, minute, second:
>   - prResolveFromDateFields
> * setting day of week:
>   - disallow, and mark deprecated
> * Date.getDate:
>   - deprecate, same as Date.localtime
> 
> I would also recommend throwing an error for Date() instead of returning localtime -- there is no obvious 'default' Date object (it could be unix epoch, now, Jan 1st 1 CE), so providing this shortcut for getting the current time is only going to confuse people. (In general: don't allow default construction of objects that have no clear default-constructed state.) localtime is long, I'd consider adding Date.now to alias it (and possibly deprecating localtime in the process, although this is useful to contrast with gmtime).

If you agree these are important and want them done, I will open a new issue for them and do them. (and if so, please let me know how I should deprecate methods -- I gather I'd need to add a new `deprecated/3.14/deprecated-3.14.sc` file to the class library and follow an example, e.g. from 3.12 with `ContiguousBlockAllocator.findPrevious`?)

Thanks,
Glen.

telephon commented:
- I wouldn't deprecate `Date.getDate`, too much hassle for the gain.
- there is a basic decision, which is kind of ambivalent still in the above: is a date always automatically resolved or only in certain places?

totalgee commented:
I think the comment is clearly saying it should always be kept internally consistent, with no explicit need to "resolve" in either direction (from DMY/hms settings, or from rawSeconds). This was a change from the original discussions. The only argument against this, I think, might be performance, but I'm not sure that's a big deal, except if someone was individually setting many individual properties. In general, in that case it would make sense just to create a new Date object.

Anyhow, I'm happy to do the work or leave it as-is, just wanted to get feedback before "throwing away" those suggestions. Thanks!

dyfer commented on SCClassLibrary/Common/Math/Date.sc:
```diff
@@ -1,13 +1,17 @@
 	*new { arg year, month, day, hour, minute, second, dayOfWeek, rawSeconds;
-		^super.newCopyArgs(year, month, day, hour, minute, second, dayOfWeek,
-				rawSeconds);
+		var instance = super.newCopyArgs(year, month, day, hour, minute, second, dayOfWeek,
+			rawSeconds);
+		^instance.prInitialize;
```
Is there a reason this is not written as 
```supercollider
super.newCopyArgs(year, month, day, hour, minute, second, dayOfWeek, rawSeconds).prInitialize;
```

dyfer commented on SCClassLibrary/Common/Math/Date.sc:
```diff
@@ -1,13 +1,17 @@
+		^instance.prInitialize;
+	}
+	*fromString { arg string, format;
+		^this.new.prFromString(string, format)
```
Does this first resolve the date into current time and then sets if from a string? If so, then this would be one place to use the proposed `*basicNew` constructor
